include/deal.II/matrix_free/fe_evaluation.h

```diff
@@ -2923,9 +2923,22 @@ namespace internal
     typedef VectorType BaseVectorType;
 
     static BaseVectorType *get_vector_component (VectorType &vec,
-                                                 const unsigned int)
+                                                 const unsigned int component)
     {
-      return &vec;
+      // FEEvaluation allows to combine several vectors from a scalar
+      // FiniteElement into a "vector-valued" FEEvaluation object with
+      // multiple components. These components can be extracted with the other
+      // get_vector_component functions. If we do not get a vector of vectors
+      // (std::vector<VectorType>, std::vector<VectorType*>, BlockVector), we
+      // must make sure that we do not duplicate the components in input
+      // and/or duplicate the resulting integrals. In such a case, we should
+      // only get the zeroth component in the vector contained set nullptr for
+      // the others which allows us to catch unintended use in
+      // read_write_operation.
+      if (component == 0)
+        return &vec;
+      else
+        return nullptr;
```

Comment:
Can't you already assert `component==0`here?

Reply:
I could, but then the user gets not context and needs to attach a debugger to see the call hierarchy. The current call at least shows `FEEvaluation::read_write_operation<VectorReader>` or something like that, so it is easier to track back in my opinion.

Reply:
Note I indeed want to have the `else` case returning a `nullptr` so the release mode bails out with a segfault rather than silently producing wrong results as we had before.

Reply:
looks like `BlockVectorSelector` is an internal class, so i think it's perfectly fine to thrown an error outside where users can have a better idea of what's going on.

Reply:
Actually, we can't put an `Assert` here since then the case in which we are using a non-block vector holding all the components of a vector-valued `FiniteElement` doesn't work anymore.
In particular, this would break
 - matrix_free/copy_feevaluation
 - matrix_free/inverse_mass_02
 - matrix_free/inverse_mass_03
 - matrix_free/matrix_vector_16
 - matrix_free/matrix_vector_stokes_base
 - matrix_free/matrix_vector_stokes_noflux
 - matrix_free/matrix_vector_stokes_notempl

In this case, we are only accessing the first component of `src`, but are initializing the other components anyway with something (previously with the same src[0], now with `nullptr`). This would suggest to call `get_vector_component` depending on the branch we are taking in `read_write_operation`... 

Since this is going to change anyway soon, I guess this is fine for the moment.

Hopefully, we can clean up the logic regarding the number of components of `FEEvaluation`, the number of components for the `FiniteElement` and the number of (used) components for `src` when merging the DG branch.

Reply:
The problem with the current interface is that we support two different ways of solving systems with multiple components. The question is whether we want to continue to support it because it requires more complicated interfaces. In the beginning, we only had the `std::vector` approach for multiple components, but that does not work when one has constraints that mix different components or wants to have a single vector in solvers. So we introduced the system approach coming from a DoFHandler in 2012, I think slightly before we actually merged the matrix-free branch into deal.II mainline. The latter has some overhead because we store multiple indices, but we might be able to fix that. DG does not have that overhead. One thing I'd like to keep is the ability to group several components together in an `FEEvaluation` object because the overhead to set multiple `FEEvaluation` really does show up, at least for polynomial degrees 1-2 in 3D and 1-4 in 2D. I guess we merge this for now and discuss the design more thoroughly when looking at the DG code.